buildSrc/src/main/java/org/elasticsearch/gradle/precommit/ForbiddenPatternsTask.java

```diff
@@ -0,0 +1,165 @@
+        getInputs().property("rules", patterns);
+    }
+
+    @InputFiles
+    @SkipWhenEmpty
+    public FileCollection files() {
+        return getProject().getConvention().getPlugin(JavaPluginConvention.class).getSourceSets()
+            .stream()
+            .map(sourceSet -> sourceSet.getAllSource().matching(filesFilter))
+            .reduce(FileTree::plus)
+            .orElse(getProject().files().getAsFileTree());
+    }
+
+    @TaskAction
+    public void checkInvalidPatterns() throws IOException {
+        Pattern allPatterns = Pattern.compile("(" + String.join(")|(", patterns.values()) + ")");
+        List<String> failures = new ArrayList<>();
+        for (File f : files()) {
+            List<String> lines = getLines(f.toPath());
+            List<Integer> invalidLines = IntStream.range(0, lines.size())
+                .filter(i -> allPatterns.matcher(lines.get(i)).find())
+                .boxed()
+                .collect(Collectors.toList());
+
+            String path = getProject().getRootProject().getProjectDir().toURI().relativize(f.toURI()).toString();
+            failures = invalidLines.stream()
+                .map(l -> new AbstractMap.SimpleEntry<>(l+1, lines.get(l)))
+                .flatMap(kv -> patterns.entrySet().stream()
+                    .filter(p -> Pattern.compile(p.getValue()).matcher(kv.getValue()).find())
+                    .map(p -> "- " + p.getKey() + " on line " + kv.getKey() + " of " + path)
+                )
+                .collect(Collectors.toList());
+        }
+        if (failures.isEmpty() == false) {
+            throw new GradleException("Found invalid patterns:\n" + String.join("\n", failures));
+        }
+
+        outputMarker.getParentFile().mkdirs();
+        Files.write(outputMarker.toPath(), "done".getBytes(StandardCharsets.UTF_8));
+    }
+
+    @OutputFile
+    public File getOutputMarker() {
+        return outputMarker;
+    }
+
+    public void exclude(String... excludes) {
+        filesFilter.exclude(excludes);
+    }
+
+    public void rule(Map<String,String> props) {
+        String name = props.remove("name");
+        if (name == null) {
+            throw new InvalidUserDataException("Missing [name] for invalid pattern rule");
+        }
+        String pattern = props.remove("pattern");
+        if (pattern == null) {
+            throw new InvalidUserDataException("Missing [pattern] for invalid pattern rule");
+        }
+        if (props.isEmpty() == false) {
+            throw new InvalidUserDataException("Unknown arguments for ForbiddenPatterns rule mapping: "
+                + props.keySet().toString());
+        }
+        // TODO: fail if pattern contains a newline, it won't work (currently)
+        patterns.put(name, pattern);
+    }
+
+    private List<String> getLines(Path path) throws IOException {
+        for (Charset encoding : Arrays.asList(StandardCharsets.UTF_8, StandardCharsets.ISO_8859_1)) {
```

Comment:
Yes, at least one of the specific files I ran into issues with was the `hunspell/en_US/en_US.dic` which wasn't readable with UTF-8.

I'm not certain how Groovy impl handled this because it used `file.eachLine('UTF-8') {...}` to go over each line in the source files. I'm wondering maybe if Groovy's `f.eachLine('UTF-8') {}` method swallows and warns in these kinds of cases?

Reply:
It would be better to exclude "*.dic" files for that project.